### ISSUE TITLE
[css-values-4] Abandon mix()?

### ISSUE THREAD

**Issue by andruud** [w3c/csswg-drafts#9343]
From https://github.com/w3c/csswg-drafts/issues/6245#issuecomment-930355155:

> &lt;TabAtkins> smfr: Would this be like a calc()?
&lt;TabAtkins> fantasai: Like, but wider.
&lt;fantasai> fantasai: It has to be able to interpolate every possible computed value in the entire space of CSS
&lt;TabAtkins> smfr: **It requires UAs to have a parallel version of calc trees, for every possible value**
&lt;TabAtkins> fantasai: You kinda already have that since everything can interp
&lt;TabAtkins> fantasai: Like, how do you interp between currentcolor and blue? No way to represent that right now. (color-mix() is coming, but this is a wider issue)
&lt;TabAtkins> fantasai: So we have lots of places where we want to interp things that don't have intermediate values
&lt;TabAtkins> smfr: That makes sense, we also invented cross-fade() to hit the image case
&lt;TabAtkins> smfr: **I'd like to hear from other impls about their thoughts on impl complexity**, and whether it makes sense to think of it in terms of calc()

(cc @smfr)

If my understanding is correct, `mix()` as currently defined is asking for way too much for not nearly enough benefit.

> It has to be able to interpolate every possible computed value in the entire space of CSS

This would indeed require that _any_ computed value that we currently store in a space-efficient manner now needs to deal with the possibility of being a _function_ instead, which then needs to be interpreted further used-value time (@bfgeek). This is _probably_ a non-starter, but it also doesn't seem that necessary. E.g. do we really need to be able to represent `display: mix(inline; block; 50%)` as a computed value when no meaningful mix between inline/block exists (I hope)?

The current path we're on elsewhere in CSS, with type-specific mix functions (`color-mix`, `font-palette-mix`) seems way more sensible:

 - There are no issues with "must be sole value" or other weirdness.
   - https://github.com/w3c/csswg-drafts/issues/9341
   - https://github.com/w3c/csswg-drafts/issues/9342
 - They can support detailed interpolation behavior (e.g. interpolation a given color space). Only a matter of time until we run into "interpolate HOW" if everything uses `mix()`.
 - We only grant mixing capabilities where it actually matters and has meaning.
 - It's possible to implement and ship gradually.

So ideally we should just abandon `mix()`. We can probably find other ways of doing everything we want in https://github.com/w3c/csswg-drafts/issues/6245, e.g. extend all _typed_ `*-mix()` functions with the capability of pulling progress from a named timeline, and things like that.

**Comment by mirisuzanne** [comment]
Abandon, or narrow in scope? What happens if this is limited to properties that can have clearly defined interpolations for transitioning/animations/etc? Are there other issues we would need to address?

**Comment by andruud** [comment]
> Abandon, or narrow in scope?

Abandon _in its current form_. :-)

If we make `mix()` more eager to resolve itself by computed-value time, then it could be workable.

> What happens if this is limited to properties that can have clearly defined interpolations for transitioning/animations/etc?

Yes. We could allow it to be _specified_ for anything, but resolve it by computed-value time if there's no interpolation behavior. Examples:

 - `display: mix(inline; block; 10%)` => computed value: `inline`. (Has no interpolation behavior).
 - `opacity: mix(0; 1; 10%)` => computed value: `0.1`. (Has interpolation behavior, but still resolves computed-value time).

And then we allow `mix()` to actually be the computed value in a select few cases, like [`<transform-list>` edge cases](https://github.com/w3c/css-houdini-drafts/issues/425), or [animating font-palettes](https://github.com/w3c/csswg-drafts/issues/8922) (if we don't go with `font-palette-mix()`). But specifications allow this on a case-by-case basis, not as a blanket "anything can mix".

**Comment by tabatkins** [comment]
The intention of mix() is that it represented *exactly* the behavior that you get from animating between the two input values; that is, if you could write `@keyframes foo { from { prop: start; } to { prop: end; }}`, then you could write `prop: mix(start; end; X%)` and get the exact same value.

This had exactly one purpose: to allow representing the computed value of an animated property that cannot be fully resolved at computed-value time, like a `transform` that requires layout info (such as `translate(50%)`) but also triggers matrix interpolation (which can only express fixed pixel values). Anything else it did was secondary; there's some minor usefulness in being able to get an interpolated value without having to interpolate manually, but that wasn't the point.

> We could allow it to be specified for anything, but resolve it by computed-value time if there's no interpolation behavior.

Good news: the spec already says *exactly* this:

> If the two `<declaration-value>`s in mix() are interpolable (without using mix() itself) as values for the property in which it is specified, the computed value of mix() is the result of interpolating these two values to the progress given by the `<percentage>`. Otherwise, the computed value of mix() is the mix() functional notation itself with its `<percentage>` computed and both `<declaration-value>`s computed as values for this property.

So outside of the very few cases that can't be represented at computed-value time, mix() already disappears at computed-value time. This shouldn't require any changes to computed values for any properties that wouldn't already require this kind of thing anyway.

That said, we do have several functions which are essentially specialized mix() versions: color-mix(), palette-mix(), cross-fade(). It *is* possible for us to just add more specialized functions as needed (a `transform-mix()`?), if we'd prefer to do that instead of having a generic facility. As you point out, a specialized function allows for more specific control when necessary (like specifying color space in the color mixing functions). I don't have a strong opinion either way, actually.

**Comment by andruud** [comment]
> Good news: the spec already says exactly this: [...]

Are you sure? Wouldn't the current spec cause e.g. the computed value of `float: mix(left; right; 10%)` to literally be `mix(left; right; 10%)`?

> If the two <declaration-value>s in mix() are interpolable (without using mix() itself) as values for the property in which it is specified, the computed value of mix() is the result of interpolating these two values to the progress given by the <percentage>. 

`left` and `right` are not interpolable, so this does not apply, which means that we fall into:

> Otherwise, the **computed value of mix() is the mix() functional notation itself** with its <percentage> computed and both <declaration-value>s computed as values for this property.

So we end up with `mix(left; right; 10%)`. This is not a situation that can happen with with regular animations. Yes, allowing this in a few problematic cases was the point of adding this in the first place (again: https://github.com/w3c/css-houdini-drafts/issues/425), so we'd of course have to let `mix()` be the computed value in those cases. But I allowing it _genreally for everything_ is going to far, as we'd need to be able to represent lots of expensive computed values internally (`mix(left; right; 10%)`) for no practical benefit.

> That said, we do have several functions which are essentially specialized mix() versions: color-mix(), palette-mix(), cross-fade(). It is possible for us to just add more specialized functions as needed (a transform-mix()?)

The good thing about `color-mix()` is that it isn't that special. It's just another thing allowed by `<color>`. This means `color-mix()` can be used wherever `<color>` goes, be it top-level or other places. So I do think `transform-mix()` and generally `specific-thing-mix()` is a better approach, _even_ if they don't offer any special interpolation behavior.

That said, if we still really want `mix()`, we can make it more approachable by resolving it computed-value time by default (also for discrete animations).

**Comment by tabatkins** [comment]
> left and right are not interpolable

`left` and `right` are definitely interpolable, they're just *discretely* interpolable; you can write keyframes that go from `left` to `right` and get a value that flips between them. `mix()` would return that value. If this was the major misunderstanding, that makes this issue make a *lot* more sense to me.

I suppose I just didn't properly handle the case where values genuinely aren't interpolable, like for `display`. That wasn't an intentional omission; I guess it makes the most sense to make that invalid.

> [specialized functions represent a type]

Yeah that's valid. Let's ask the larger WG about this; I'm happy to go in either direction.

(But if we drop `mix()` we do need `transform-mix()`.)

**Comment by Loirooriol** [comment]
I guess [discrete interpolation](https://drafts.csswg.org/web-animations-1/#discrete) is still an interpolation, so `float: mix(left; right; 10%)` would compute to `float: left`?

And can the broken "<s" in the spec be fixed?

**Comment by Loirooriol** [comment]
@tabatkins I think #6429 made `display` animatable. Examples of non-animatable properties would be the animation properties: https://drafts.csswg.org/css-animations/#animation-definition

**Comment by tabatkins** [comment]
Ah and also I already *do* specify that it's invalid if the values aren't interpolable.

**Comment by andruud** [comment]
> If this was the major misunderstanding, that makes this issue make a lot more sense to me.

Ah, OK. I think that was it. See also the IRC discussion in the first post. These comments reinforced (or caused?) my misunderstanding:

> fantasai: It has to be able to interpolate every possible computed value in the entire space of CSS

> smfr: It requires UAs to have a parallel version of calc trees, for every possible value

**Comment by tabatkins** [comment]
I added some more examples calling out these cases.

But your point about the specialized functions being a useful type is still valid, so let's still talk about this in the group.

**Comment by Loirooriol** [comment]
I suggested moving it to L5 in https://github.com/w3c/csswg-drafts/issues/6753#issuecomment-954142319, but @fantasai said

> No, we need that one [`mix()`] to handle serialization of interpolation of transforms.

**Comment by tabatkins** [comment]
Yeah, but that's been an issue for years, so we're not making things worse in the interim. ^_^

**Comment by tabatkins** [comment]
We've added `transform-mix()` to Values 5, so I *think* all the values that can't be represented now have a representation.

**Comment by LeaVerou** [comment]
I think the goal of `mix()` is a little murky and if we're keeping it we need to reach consensus on that:
Is it intended to…
1. Allow authors to tap into the default interpolation algorithm without having to resort to hacks like paused animations with negative delays or complex logic, i.e. any `mix()` value always has an equivalent CSS value without `mix()`, essentially a higher level abstraction over existing syntax.
2. Or to *also* serve as the (potentially sole) representation of the interpolated value for certain types, essentially reducing the need to add more specific `*-mix()` functions to CSS?

These involve different tradeoffs and different design decisions. I suspect 1 is easier to implement (?) and from the author's perspective, 1 covers their use cases just fine. 2 does make for a smaller API surface, but its benefits are largely internal.

From an author perspective, it’s entirely unclear why `mix()` needs to be a whole value, or how exactly does this work. E.g. I imagine most authors would not be able to tell with reasonable confidence whether this would be allowed:

```css
--untyped-custom-property: mix(...);
property: something var(--untyped-custom-property), something else;
```

> The current path we're on elsewhere in CSS, with type-specific mix functions (color-mix, font-palette-mix) seems way more sensible

`mix()` covers distinct use cases over these.

---

I wonder if a reasonable MVP would be to only ship the [`mix(<progress> of <animation-name>)`](https://drafts.csswg.org/css-values-5/#mix) syntax. It is unclear from the current spec what property that returns, but it seems like that should be a parameter that would default to the current property:

```
<mix()> = mix( [ <custom-ident> at ]? <progress> of <animation-name> )
```

This already makes the current workarounds *a lot* easier, doesn’t have warts like having to be the whole value of a property, and can even be generalized in a limited way like so:

```css
@keyframes mix-helper {
	from { background: var(--from);
	to { background: var(--to);
}

.foo {
	--from: yellow;
	--to: linear-gradient(in oklch, yellow, oklch(.5 .2 180));
	other-property: mix(background at 30% of mix-helper);
}
```

**Comment by Loirooriol** [comment]
In `aspect-ratio`, `mix()` allows the nice logarithmic interpolation of ratios, e.g.

```css
aspect-ratio: mix(50%, 1, 9); /* 3 */
```

If I use `aspect-ratio: calc-mix(50%, 1, 9)` I will get `5` instead.

I'm not sure whether a `ratio-mix()` is worth it, but this is the use-case that comes to mind.

**Comment by Crissov** [comment]
Isn’t this particular use case of `ratio-mix()` simply the geometric mean?

**Comment by Loirooriol** [comment]
Yes, specifically at 50% it can be simplified as a basic geometric mean, but otherwise it's a weighted geometric mean and you need something like the formulas above.

**Comment by tabatkins** [comment]
> In aspect-ratio, mix() allows the nice logarithmic interpolation of ratios

Sure, but (a) the interpolated result is completely expressible as a number, and (b) you can reproduce it by hand with `log()` and `exp()` functions: `aspect-ratio: exp(log(1) * p + log(9) * (1 - p))` (or `exp(calc-mix(p, log(1), log(9)))` if you want).

If we think that exposing a *convenient shorthand* for logarithmic interpolation would be useful, we can do so, but there's nothing *requiring* us to do so like the other cases we've added a *-mix() for. (Well, `calc-mix()` is just a convenience shorthand too, but it justified itself well.)
